// mqtt_service.proto - MQTT 消息代理服务 gRPC 接口
// 提供统一的 MQTT 消息发布/订阅接口，支持多租户和设备管理
syntax = "proto3";

package isa.mqtt;

option go_package = "github.com/isa-cloud/isa_cloud/pkg/proto/mqtt";

import "common.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ========================================
// MQTT 消息代理服务
// ========================================
// 这是一个统一的 MQTT 消息服务，为所有后端服务提供设备通信能力
// 支持：
// - 用户和组织级别的主题隔离
// - 设备注册和管理
// - 消息发布和订阅
// - QoS 质量保证
// - 保留消息和遗嘱消息
// - 主题通配符订阅

service MQTTService {
    // ========== 连接管理 ==========
    rpc Connect(ConnectRequest) returns (ConnectResponse);
    rpc Disconnect(DisconnectRequest) returns (DisconnectResponse);
    rpc GetConnectionStatus(ConnectionStatusRequest) returns (ConnectionStatusResponse);
    
    // ========== 消息发布 ==========
    // 发布单条消息
    rpc Publish(PublishRequest) returns (PublishResponse);
    // 批量发布消息
    rpc PublishBatch(PublishBatchRequest) returns (PublishBatchResponse);
    // 发布 JSON 消息（自动序列化）
    rpc PublishJSON(PublishJSONRequest) returns (PublishResponse);
    
    // ========== 消息订阅 ==========
    // 订阅主题（流式接收消息）
    rpc Subscribe(SubscribeRequest) returns (stream MessageResponse);
    // 批量订阅多个主题
    rpc SubscribeMultiple(SubscribeMultipleRequest) returns (stream MessageResponse);
    // 取消订阅
    rpc Unsubscribe(UnsubscribeRequest) returns (UnsubscribeResponse);
    // 获取订阅列表
    rpc ListSubscriptions(ListSubscriptionsRequest) returns (ListSubscriptionsResponse);
    
    // ========== 设备管理 ==========
    rpc RegisterDevice(RegisterDeviceRequest) returns (RegisterDeviceResponse);
    rpc UnregisterDevice(UnregisterDeviceRequest) returns (UnregisterDeviceResponse);
    rpc ListDevices(ListDevicesRequest) returns (ListDevicesResponse);
    rpc GetDeviceInfo(GetDeviceInfoRequest) returns (GetDeviceInfoResponse);
    rpc UpdateDeviceStatus(UpdateDeviceStatusRequest) returns (UpdateDeviceStatusResponse);
    
    // ========== 主题管理 ==========
    rpc GetTopicInfo(GetTopicInfoRequest) returns (GetTopicInfoResponse);
    rpc ListTopics(ListTopicsRequest) returns (ListTopicsResponse);
    rpc ValidateTopic(ValidateTopicRequest) returns (ValidateTopicResponse);
    
    // ========== 保留消息 ==========
    rpc SetRetainedMessage(SetRetainedMessageRequest) returns (SetRetainedMessageResponse);
    rpc GetRetainedMessage(GetRetainedMessageRequest) returns (GetRetainedMessageResponse);
    rpc DeleteRetainedMessage(DeleteRetainedMessageRequest) returns (DeleteRetainedMessageResponse);
    
    // ========== 统计和监控 ==========
    rpc GetStatistics(GetStatisticsRequest) returns (GetStatisticsResponse);
    rpc GetDeviceMetrics(GetDeviceMetricsRequest) returns (GetDeviceMetricsResponse);
    
    // ========== 健康检查 ==========
    rpc HealthCheck(MQTTHealthCheckRequest) returns (MQTTHealthCheckResponse);
}

// ========================================
// 数据结构
// ========================================

// QoS 级别
enum QoSLevel {
    QOS_AT_MOST_ONCE = 0;   // 最多一次（不保证送达）
    QOS_AT_LEAST_ONCE = 1;  // 至少一次（可能重复）
    QOS_EXACTLY_ONCE = 2;   // 恰好一次（保证送达且不重复）
}

// 设备状态
enum DeviceStatus {
    DEVICE_UNKNOWN = 0;     // 未知
    DEVICE_ONLINE = 1;      // 在线
    DEVICE_OFFLINE = 2;     // 离线
    DEVICE_SLEEPING = 3;    // 休眠
    DEVICE_ERROR = 4;       // 错误
}

// 设备信息
message DeviceInfo {
    string device_id = 1;                           // 设备 ID
    string device_name = 2;                         // 设备名称
    string device_type = 3;                         // 设备类型（sensor, actuator, gateway）
    string user_id = 4;                             // 所属用户
    string organization_id = 5;                     // 所属组织
    DeviceStatus status = 6;                        // 设备状态
    google.protobuf.Timestamp registered_at = 7;    // 注册时间
    google.protobuf.Timestamp last_seen = 8;        // 最后在线时间
    map<string, string> metadata = 9;               // 设备元数据
    repeated string subscribed_topics = 10;         // 订阅的主题列表
    int64 messages_sent = 11;                       // 发送消息数
    int64 messages_received = 12;                   // 接收消息数
}

// 主题信息
message TopicInfo {
    string topic = 1;                               // 主题名称
    int64 subscriber_count = 2;                     // 订阅者数量
    int64 message_count = 3;                        // 消息总数
    google.protobuf.Timestamp last_message_time = 4; // 最后消息时间
    bool has_retained_message = 5;                  // 是否有保留消息
    string user_id = 6;                             // 所属用户（隔离用）
    string organization_id = 7;                     // 所属组织
}

// 消息体
message MQTTMessage {
    string topic = 1;                               // 主题
    bytes payload = 2;                              // 消息内容
    QoSLevel qos = 3;                               // QoS 级别
    bool retained = 4;                              // 是否保留
    google.protobuf.Timestamp timestamp = 5;        // 时间戳
    string message_id = 6;                          // 消息 ID
    map<string, string> properties = 7;             // 消息属性（MQTT 5.0）
}

// ========================================
// 请求/响应消息
// ========================================

// 连接请求
message ConnectRequest {
    string user_id = 1;                             // 用户 ID
    string client_id = 2;                           // 客户端 ID
    string username = 3;                            // MQTT 用户名（可选）
    string password = 4;                            // MQTT 密码（可选）
    bool clean_session = 5;                         // 是否清除会话
    int32 keep_alive = 6;                           // 心跳间隔（秒）
    
    // 遗嘱消息（Last Will and Testament）
    string will_topic = 7;                          // 遗嘱主题
    bytes will_payload = 8;                         // 遗嘱内容
    QoSLevel will_qos = 9;                          // 遗嘱 QoS
    bool will_retain = 10;                          // 遗嘱保留
}

message ConnectResponse {
    bool success = 1;
    string session_id = 2;                          // 会话 ID
    string message = 3;
    bool session_present = 4;                       // 会话是否存在（clean_session=false 时）
}

// 断开连接请求
message DisconnectRequest {
    string session_id = 1;
    string user_id = 2;
}

message DisconnectResponse {
    bool success = 1;
    string message = 2;
}

// 连接状态请求
message ConnectionStatusRequest {
    string session_id = 1;
    string user_id = 2;
}

message ConnectionStatusResponse {
    bool connected = 1;
    google.protobuf.Timestamp connected_at = 2;
    int64 messages_sent = 3;
    int64 messages_received = 4;
}

// 发布消息请求
message PublishRequest {
    string user_id = 1;                             // 用户 ID（权限控制）
    string session_id = 2;                          // 会话 ID
    string topic = 3;                               // 主题
    bytes payload = 4;                              // 消息内容
    QoSLevel qos = 5;                               // QoS 级别
    bool retained = 6;                              // 是否保留
    map<string, string> properties = 7;             // 消息属性
}

message PublishResponse {
    bool success = 1;
    string message_id = 2;                          // 消息 ID
    string message = 3;
    google.protobuf.Timestamp published_at = 4;
}

// 批量发布请求
message PublishBatchRequest {
    string user_id = 1;
    string session_id = 2;
    repeated PublishRequest messages = 3;           // 消息列表
}

message PublishBatchResponse {
    bool success = 1;
    int32 published_count = 2;                      // 成功发布数量
    int32 failed_count = 3;                         // 失败数量
    repeated string message_ids = 4;                // 消息 ID 列表
    repeated string errors = 5;                     // 错误信息
}

// 发布 JSON 请求
message PublishJSONRequest {
    string user_id = 1;
    string session_id = 2;
    string topic = 3;
    google.protobuf.Struct data = 4;                // JSON 数据（自动序列化）
    QoSLevel qos = 5;
    bool retained = 6;
}

// 订阅请求
message SubscribeRequest {
    string user_id = 1;                             // 用户 ID
    string session_id = 2;                          // 会话 ID
    string topic_filter = 3;                        // 主题过滤器（支持通配符 +, #）
    QoSLevel qos = 4;                               // 请求的 QoS
}

// 批量订阅请求
message SubscribeMultipleRequest {
    string user_id = 1;
    string session_id = 2;
    repeated TopicSubscription subscriptions = 3;
}

message TopicSubscription {
    string topic_filter = 1;
    QoSLevel qos = 2;
}

// 消息响应（流式返回）
message MessageResponse {
    string topic = 1;
    bytes payload = 2;
    QoSLevel qos = 3;
    bool retained = 4;
    google.protobuf.Timestamp timestamp = 5;
    string message_id = 6;
    map<string, string> properties = 7;
}

// 取消订阅请求
message UnsubscribeRequest {
    string user_id = 1;
    string session_id = 2;
    repeated string topic_filters = 3;              // 要取消的主题列表
}

message UnsubscribeResponse {
    bool success = 1;
    int32 unsubscribed_count = 2;
    string message = 3;
}

// 订阅列表请求
message ListSubscriptionsRequest {
    string user_id = 1;
    string session_id = 2;
}

message ListSubscriptionsResponse {
    repeated TopicSubscription subscriptions = 1;
}

// 设备注册请求
message RegisterDeviceRequest {
    string user_id = 1;
    string organization_id = 2;
    string device_id = 3;
    string device_name = 4;
    string device_type = 5;
    map<string, string> metadata = 6;
}

message RegisterDeviceResponse {
    bool success = 1;
    DeviceInfo device = 2;
    string message = 3;
}

// 设备注销请求
message UnregisterDeviceRequest {
    string user_id = 1;
    string device_id = 2;
}

message UnregisterDeviceResponse {
    bool success = 1;
    string message = 2;
}

// 设备列表请求
message ListDevicesRequest {
    string user_id = 1;
    string organization_id = 2;
    DeviceStatus status = 3;                        // 过滤状态（可选）
    int32 page = 4;                                 // 页码
    int32 page_size = 5;                            // 每页数量
}

message ListDevicesResponse {
    repeated DeviceInfo devices = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// 设备信息请求
message GetDeviceInfoRequest {
    string user_id = 1;
    string device_id = 2;
}

message GetDeviceInfoResponse {
    DeviceInfo device = 1;
}

// 更新设备状态请求
message UpdateDeviceStatusRequest {
    string user_id = 1;
    string device_id = 2;
    DeviceStatus status = 3;
    map<string, string> metadata = 4;               // 更新元数据（可选）
}

message UpdateDeviceStatusResponse {
    bool success = 1;
    DeviceInfo device = 2;
}

// 主题信息请求
message GetTopicInfoRequest {
    string user_id = 1;
    string topic = 2;
}

message GetTopicInfoResponse {
    TopicInfo topic_info = 1;
}

// 主题列表请求
message ListTopicsRequest {
    string user_id = 1;
    string organization_id = 2;
    string prefix = 3;                              // 主题前缀过滤（可选）
    int32 page = 4;
    int32 page_size = 5;
}

message ListTopicsResponse {
    repeated TopicInfo topics = 1;
    int32 total_count = 2;
}

// 验证主题请求
message ValidateTopicRequest {
    string topic = 1;                               // 主题名称
    bool allow_wildcards = 2;                       // 是否允许通配符
}

message ValidateTopicResponse {
    bool valid = 1;
    string message = 2;                             // 验证消息
    repeated string violations = 3;                 // 违规项
}

// 保留消息请求
message SetRetainedMessageRequest {
    string user_id = 1;
    string topic = 2;
    bytes payload = 3;
    QoSLevel qos = 4;
}

message SetRetainedMessageResponse {
    bool success = 1;
    string message = 2;
}

message GetRetainedMessageRequest {
    string user_id = 1;
    string topic = 2;
}

message GetRetainedMessageResponse {
    bool found = 1;
    MQTTMessage message = 2;
}

message DeleteRetainedMessageRequest {
    string user_id = 1;
    string topic = 2;
}

message DeleteRetainedMessageResponse {
    bool success = 1;
    string message = 2;
}

// 统计信息请求
message GetStatisticsRequest {
    string user_id = 1;
    string organization_id = 2;
}

message GetStatisticsResponse {
    int64 total_devices = 1;                        // 设备总数
    int64 online_devices = 2;                       // 在线设备数
    int64 total_topics = 3;                         // 主题总数
    int64 total_subscriptions = 4;                  // 订阅总数
    int64 messages_sent_today = 5;                  // 今日发送消息数
    int64 messages_received_today = 6;              // 今日接收消息数
    int64 active_sessions = 7;                      // 活跃会话数
    map<string, int64> device_type_distribution = 8; // 设备类型分布
}

// 设备指标请求
message GetDeviceMetricsRequest {
    string user_id = 1;
    string device_id = 2;
    google.protobuf.Timestamp start_time = 3;
    google.protobuf.Timestamp end_time = 4;
}

message GetDeviceMetricsResponse {
    string device_id = 1;
    int64 messages_sent = 2;
    int64 messages_received = 3;
    int64 bytes_sent = 4;
    int64 bytes_received = 5;
    repeated TimeSeriesPoint message_rate = 6;      // 消息速率时序数据
    repeated TimeSeriesPoint error_rate = 7;        // 错误率时序数据
}

message TimeSeriesPoint {
    google.protobuf.Timestamp timestamp = 1;
    double value = 2;
}

// 健康检查
message MQTTHealthCheckRequest {
    bool deep_check = 1;                            // 是否进行深度检查
}

message MQTTHealthCheckResponse {
    bool healthy = 1;
    string broker_status = 2;                       // Broker 状态
    int64 active_connections = 3;                   // 活跃连接数
    google.protobuf.Timestamp checked_at = 4;
    string message = 5;
}



